Make groovy parsing of console log configurable

pjdarton opened the pull request:
[JENKINS-54832](https://issues.jenkins.io/browse/JENKINS-54832) alludes to security concerns regarding allowing (admin-configured) groovy parser code to run on the Jenkins master node, and that this is why the groovy parsers are unable to scan the build console log.
My (limited) understanding is that this is because the parser code is not (currently) run within a security-sandbox and therefore there are risks involved in allowing such code to run on the master (where the console log would be scanned).
However, a Jenkins admin _may_ decide that, as they control the parser code and can decide what code to allow in or not (just as they decide what groovy methods to permit in a sandboxed environment), they wish to allow this code to run on the master unhindered.

This PR allows such a choice to be a tick-box configurable option within the "Groovy Based Warnings Parsers" section of the Jenkins global configuration (Manage Jenkins -> Configure System):
![image](https://user-images.githubusercontent.com/984537/122971451-f8e88b00-d386-11eb-9df7-9716af149600.png)
It defaults to false (un-ticked) in order to ensure that existing users who upgrade are unaffected.

With the option un-ticked, groovy parsers missing a pattern will (now) indicate that a value is required:
![image](https://user-images.githubusercontent.com/984537/122976822-b5911b00-d38c-11eb-9ab4-58a6c5c0336b.png)
Previously the first indication was when the build threw an exception and logged `FATAL: Static analysis tool .... cannot scan console log output, please define a file pattern`

With the option ticked, groovy parsers are allowed to not specify a pattern and the web form indicates this:
![image](https://user-images.githubusercontent.com/984537/122976634-7bc01480-d38c-11eb-826c-04a25e37cbd6.png)
...and any build using the parser will have its console logged as expected instead of throwing an exception.

TL;DR: This lets the Jenkins administrator decide.

uhafner commented on plugin/src/main/java/io/jenkins/plugins/analysis/core/model/ReportScanningTool.java:
```diff
@@ -294,6 +294,9 @@ public FormValidation doCheckPattern(@AncestorInPath final AbstractProject<?, ?>
             if (!JENKINS.hasPermission(Item.CONFIGURE, project)) {
                 return FormValidation.ok();
             }
+            if ((pattern==null || pattern.trim().isEmpty()) && !canScanConsoleLog()) {
```
```suggestion
            if ((pattern==null || pattern.trim().isEmpty()) && !canScanConsoleLog()) {
```

uhafner commented on plugin/src/main/java/io/jenkins/plugins/analysis/core/model/ReportScanningTool.java:
```diff
@@ -294,6 +294,9 @@ public FormValidation doCheckPattern(@AncestorInPath final AbstractProject<?, ?>
             if (!JENKINS.hasPermission(Item.CONFIGURE, project)) {
                 return FormValidation.ok();
             }
+            if ((pattern == null || pattern.trim().isEmpty()) && !canScanConsoleLog()) {
```
This will break existing parsers that provide a default pattern like CheckStyle.

It would make sense to provide that improved check in a method like `hasPattern` (or similar).

pjdarton replied:
Apologies - I'm being a bit myopic and focussed purely on the Groovy use case...
How about:
```
            final boolean thereIsNoPatternConfigured = pattern == null || pattern.trim().isEmpty();
            final boolean thereIsNoDefaultPattern = getPattern().trim().isEmpty();
            if (thereIsNoPatternConfigured && thereIsNoDefaultPattern && !canScanConsoleLog()) {
                return FormValidation.error(Messages.ReportScanningTool_PatternIsEmptyAndConsoleParsingDisabled());
            }
```

uhafner replied:
I think you need to write a method (or variable):
```
boolean hasPattern() {
    return StringUtils.isNotEmpty(getPattern()) || StringUtils.isNotEmpty(getDescriptor().getPattern());
}
```

pjdarton replied:
Done - I added a "hasDefaultPattern()" method and used that.

uhafner commented on plugin/src/main/resources/io/jenkins/plugins/analysis/core/model/ReportScanningTool/help-pattern.html:
```diff
@@ -2,4 +2,7 @@
     setting that specifies the report files to scan for issues. Multiple patterns can be separated by space or comma.
     Note that such a pattern is resolved in Jenkins' workspace, so the paths must be relative only.
- </div>
+    <p>
+    If no pattern is specified then the build's console text will be scanned instead.
+    </p>
```
This is not correct see comment above about CheckStyle

pjdarton replied:
Hmm, yes...
I think that it would be beneficial to tell people this additional information _when it is true_ though.
I'll see if I can figure out a way of making it conditional on `getPattern()` returning an empty string - I think it's possible to have the help generated by groovy instead of simple HTML...

pjdarton replied:
I experimented and found that a help-pattern.jelly file isn't passed enough information to dynamically decide what to do.
So instead I extended the existing choose/when/otherwise pattern in the config.jelly to switch between 3 different help files as well as descriptions.
I've checked and this works for e.g. CCM, Brakeman etc as well as GroovyScript - the help, description and validation now document/enforce the same logic.

uhafner commented on plugin/src/main/resources/io/jenkins/plugins/analysis/warnings/groovy/GroovyScript/config.jelly:
```diff
@@ -5,10 +5,20 @@
+  <j:choose>
+    <j:when test="${descriptor.canScanConsoleLog()}">
+      <f:entry title="${%title.pattern}" field="pattern"
+               description="${%description.consoleOrPattern}">
```
```suggestion
               description="${%description.consoleOrPattern('http://ant.apache.org/manual/Types/fileset.html')}">
```

uhafner commented on plugin/src/main/resources/io/jenkins/plugins/analysis/warnings/groovy/GroovyScript/config.jelly:
```diff
@@ -5,10 +5,20 @@
+    </j:when>
+    <j:otherwise>
+      <f:entry title="${%title.pattern}" field="pattern"
+               description="${%description.pattern}">
```
```suggestion
               description="${%description.pattern('http://ant.apache.org/manual/Types/fileset.html')}">
```

uhafner commented on plugin/src/main/resources/io/jenkins/plugins/analysis/warnings/groovy/GroovyScript/help-pattern.html:
```diff
@@ -1 +0,0 @@
-Note that due to security reasons scanning the console log using a Groovy script is not allowed.
```
Rather than deleting this file it would make sense to copy the help from the other `help-pattern.html` file.

pjdarton replied:
Ideally, I'd like the other file to work for both cases 😉 
However, if I can't make it do that then, yes, I'll make a GroovyScript-specific version here.

uhafner replied:
I tried that but it seems that Jenkins core does not support inheritance on those files. So I ended up copying the same files to all the folders.

pjdarton replied:
I've experimented and I've found that, with the GroovyScript's help-pattern.html gone, Jenkins does inherit.
So I've made the inherited one flexible and able to cope with all 3 scenarios (default pattern, optional pattern, mandatory pattern) and I've checked that this works on e.g. both GroovyScript, CCM and others.

I did find that if you define a config.jelly file then that'll be used _instead_ of the parent's config.jelly but it's able to reference help files in the parent without a problem.

uhafner commented on plugin/src/main/resources/io/jenkins/plugins/analysis/warnings/groovy/ParserConfiguration/config.jelly:
```diff
@@ -7,10 +7,13 @@
         <f:entry title="${%Parsers}" description="${%title.description}">
           <i:repeatable field="parsers"/>
         </f:entry>
+        <f:entry field="consoleLogScanningPermitted" title="${%Permit Parsing of Console?}" description="${%consoleLogScanningPermitted.description}">
```
Can you please use a property here as well for the `title`, I'm not sure how the question mark works in properties files.

uhafner commented on plugin/src/main/resources/io/jenkins/plugins/analysis/warnings/groovy/ParserConfiguration/config.properties:
```diff
@@ -1 +1,2 @@
 title.description=Create new parsers for the warnings plug-in based on a regular expression and a Groovy script.
+consoleLogScanningPermitted.description=Allow these parsers to act upon the build console.
```
Jenkins's security team mentioned that it would be helpful to make the text from your help file visible here as well (`help-canScanConsoleLog.html`). At least that checking the box can introduce a (potential) security risk.

pjdarton replied:
I'd prefer not to clutter the config page with a duplication of the help info.
...but I can do better than that - I'll add a doCheck method that'll show a warning when the checkbox is ticked that tells people to go read the help.
I suspect that a yellow warning will get more attention than static text.

uhafner commented on plugin/src/main/resources/io/jenkins/plugins/analysis/warnings/groovy/ParserConfiguration/help-canScanConsoleLog.html:
```diff
@@ -0,0 +1,14 @@
+<div>
+  <p>
+    Console Log parsing has to be processed directly on the Jenkins master
```
```suggestion
    Console Log parsing has to be processed directly on the Jenkins controller
```

uhafner commented on plugin/src/main/resources/io/jenkins/plugins/analysis/warnings/groovy/ParserConfiguration/help-canScanConsoleLog.html:
```diff
@@ -0,0 +1,14 @@
+<div>
+  <p>
+    Console Log parsing has to be processed directly on the Jenkins master
+    which means running the parser groovy code (configured here) directly on the master which <em>CAN BE UNSAFE</em> from a security point of view.
```
```suggestion
    which means running the parser groovy code (configured here) directly on the controller which <em>CAN BE UNSAFE</em> from a security point of view.
```

uhafner commented on plugin/src/main/resources/io/jenkins/plugins/analysis/warnings/groovy/ParserConfiguration/help-canScanConsoleLog.html:
```diff
@@ -0,0 +1,14 @@
+    If you do not accept this risk then you should not enable this feature
+    ...
+    but if you do not enable this feature then the groovy parsers that you define will only be able to parse files in the build workspace
+    (safely processed on the nodes responsible for the builds rather than natively on the master)
```
```suggestion
    (safely processed on the agents responsible for the builds rather than natively on the controller)
```

uhafner reviewed (changes requested):
Thanks for adding that enhancement! 

Can you please add a small test that shows that scanning the console log on the controller really works (a copy of `shouldRemoveConsoleLogNotesBeforeRemovingColorCodes` for instance with the Groovy Pep8 parser)?

pjdarton commented:
I believe I've addressed all the issues _except_ adding a "small test that shows that scanning the console log".
I'm not sure how to do that as, in my IDE, `StepsITest` is riddled with compilation errors because it can't `import static io.jenkins.plugins.analysis.core.assertions.Assertions.*` and, despite some searching, I can't find where that package/class is defined either.
...and if I do `mvn install` (or even `mvn '-Dtest=StepsITest#shouldShowWarningsOfGroovyParser' test`) on the command-line then it fails for other reasons (after a lengthy delay, probably because it doesn't like running on my Linux VM whose working directory is a Windows filesystem).
i.e. My dev environment isn't up to this kind of task and I'm not sure how to fix it.

What I'd like to do w.r.t. a unit-test is something like...
- Rename `shouldShowWarningsOfGroovyParser` to `shouldShowWarningsOfGroovyParserWhenScanningFileInWorkspace`
- Create a new test `shouldShowWarningsOfGroovyParserWhenScanningConsoleLogWhenThatIsPermitted`.
Read in `"pep8Test.txt"` as a series of lines, turn each into an `echo` statement to go into the `asStage` data before the `scanForIssues` part.
Have the scan specify `pattern:''` so it'll only look at the console.
Call `configuration.setConsoleLogScanningPermitted(true)` so this is permitted.
Run the build and expect success.
Expect to get the same scan results as for the original `shouldShowWarningsOfGroovyParser` test.
- Create another test called `shouldFailUsingGroovyParserToScanConsoleLogWhenThatIsForbidden` that's nearly identical to the previous one but which does NOT call `configuration.setConsoleLogScanningPermitted(true)` and which expects the build to fail (with no scanned results).

However, because I've got issues with my development environment, I'd be mostly "coding blind" with a 45-minute turnaround time between "commit & push" and seeing the results which just isn't viable; I think I'm going to struggle to get any non-trivial unit-test working without help 😞

uhafner commented:
Which ID are you using? In IntelliJ you sometimes need to manually add the folder `target/generated-test-sources/assertj-assertions` (see https://github.com/uhafner/warnings-ng-plugin-devenv#installation-steps) as test source folders.

pjdarton commented:
Thanks for that hint - that sorted it and got me going again 😁 

FYI there's no (obvious) link to the "Warnings Plugin - Development Environment" from the main README.md file.
I've (now) spotted that there's a link from the contributions page ... but I didn't read that page before 😞 
Might be an idea to link to the Dev-env and/or contributions page from the main README.md file, e.g. in the "If your tool is not yet supported" paragraph so the folks have more of an opportunity to spot it (or to feel more foolish about not having seen it 😉)

pjdarton commented on plugin/src/main/java/io/jenkins/plugins/analysis/warnings/groovy/ParserConfiguration.java:
```diff
@@ -84,6 +87,50 @@ public void setParsers(final List<GroovyParser> parsers) {
+    // Similarly, there's no need to permission-check either.
+    // If it is ever changed to do anything non-trivial that might leak secrets then
+    // it must be changed to @POST and checkMethod="post" be added to the checkbox.
+    // There is a matching entry in archunit_ignore_patterns.txt to permit this.
```
Note: I did try adding checkMethod="post" to the checkbox element, but it didn't work - it looks like checkbox does not support that attribute (and there's no safe-checkbox either).
i.e. right now, we have no other option but to use a non-POST callback ... but that's OK because this doCheck method does not leak data (you either get a fxied-text warning, or you get no error - there's no side-effects).

pjdarton commented on plugin/src/main/resources/io/jenkins/plugins/analysis/warnings/groovy/ParserConfiguration/config.jelly:
```diff
@@ -7,10 +7,13 @@
       <j:otherwise>
         <f:block>
-          You have no permission to edit this section (required permission: RUN_SCRIPTS).
+          You have no permission to edit this section (required permission: ADMINISTER).
```
FYI the underlying code checks for Jenkins.ADMINISTER - there's no check for RUN_SCRIPTS.

uhafner replied:
Yes, this has been replaced in Jenkins core recently and I forgot to change the documentation as well.

pjdarton commented on plugin/src/main/resources/io/jenkins/plugins/analysis/warnings/groovy/ParserConfiguration/help-consoleLogScanningPermitted.html:
```diff
@@ -0,0 +1,14 @@
+    but if you do not enable this feature then the groovy parsers that you define will only be able to parse files in the build workspace
+    (safely processed on the agents responsible for the builds rather than natively on the controller)
+    rather than the build console log itself.
+  </p>
```
This text should be reviewed by someone who's 100% aware of the Jenkins-security-team's concerns about this.

uhafner replied:
@daniel-beck @Wadeck Is this ok for you?

daniel-beck replied:
As I wrote elsewhere, I think we need to revisit the security considerations for Overall/Administer users after we retired the RunScripts permission around Jenkins 2.222.

That said, I recommend you consider whether you want to allow users with Overall/Manage permission (JEP-223) to configure these parsers at some point in the near future.

FWIW the real long term solution for all of this seems to be Script Security integration anyway. Then none of this is a concern -- an admin will not be impacted (their scripts will be pre-approved), Overall/Manage users' parsers need to be approved by admins first or sandboxes, and the scripts can run wherever.

pjdarton replied:
Yes  I view this PR as only being useful until this plug-in has sandboxed scripts.
(and if I hadn't been in a big hurry to "just get something working now!" then I would've looked into that instead).
Once scripts are sandboxed then this configuration option is irrelevant and can be hard coded as "on".

uhafner replied:
I implemented the script security integration already some years ago. I removed that part because then the parsers did not work on the agent anymore. Script security is only available on the controller side.

pjdarton commented on plugin/src/test/java/io/jenkins/plugins/analysis/warnings/groovy/ParserConfigurationTest.java:
```diff
@@ -19,14 +20,14 @@
 
     @Test
-    void shouldHaveNoRootFoldersWhenCreated() {
+    void shouldHaveNoParsersWhenCreated() {
```
I've renamed this test case to better describe what it's testing.
(I suspect there was some copy/paste of test names when this class was originally created)

pjdarton commented on plugin/src/test/java/io/jenkins/plugins/analysis/warnings/groovy/ParserConfigurationTest.java:
```diff
@@ -19,14 +20,14 @@
 
     @Test
-    void shouldSaveConfigurationIfFoldersAreAdded() {
+    void shouldSaveConfigurationIfParsersAreAdded() {
```
I've renamed this test case to better describe what it's testing.
(I suspect there was some copy/paste of test names when this class was originally created)

pjdarton commented on plugin/src/test/java/io/jenkins/plugins/analysis/warnings/steps/StepsITest.java:
```diff
@@ -732,6 +734,63 @@ public void shouldShowWarningsOfGroovyParser() {
+        job.setDefinition(asStage(stages.toArray(new String[0])));
+
+        ParserConfiguration configuration = ParserConfiguration.getInstance();
+        configuration.setConsoleLogScanningPermitted(true);
```
CPD is complaining that it `Found duplicated code` and I don't intend to fix that - many unit tests are fairly repetitive by nature.
I've extracted some code out into common methods to reduce duplication but there will inevitably be some commonality between related tests.

uhafner replied:
Yes, this is ok! There is no quality gate enforced for CPD anyway, it is just to provide some feedback.

pjdarton commented on plugin/src/test/resources/archunit_ignore_patterns.txt:
```diff
@@ -1 +1,6 @@
+# ParserConfiguration.doCheckConsoleLogScanningPermitted does not require @POST or a permission check.
+# As-implemented, it does not leak data, even if called illegally.
+.*ParserConfiguration\.doCheckConsoleLogScanningPermitted.*POST.*
+.*Permission.*ParserConfiguration\.doCheckConsoleLogScanningPermitted.*
```
I tried to avoid having to add any ignores, but there's no reason why that specific doCheck method requires a POST - that's only needed when a doCheck method can leak information.

pjdarton commented:
OK, I think this is all sorted now; I belive I've addressed all the previous review comments and sorted all the CPD/CheckStyle/PMD/coverage issues too.

uhafner commented:
Thanks! I'll hopefully find time for a review this evening.

uhafner commented:
> FYI there's no (obvious) link to the "Warnings Plugin - Development Environment" from the main README.md file.
I've (now) spotted that there's a link from the contributions page ... but I didn't read that page before 😞
Might be an idea to link to the Dev-env and/or contributions page from the main README.md file, e.g. in the "If your tool is not yet supported" paragraph so the folks have more of an opportunity to spot it (or to feel more foolish about not having seen it 😉

Yes, I need to add a section to the README. Thanks for pointing that out.

uhafner commented on plugin/src/main/resources/io/jenkins/plugins/analysis/warnings/groovy/ParserConfiguration/help-consoleLogScanningPermitted.html:
```diff
@@ -0,0 +1,14 @@
+    Console Log parsing has to be processed directly on the Jenkins controller
+    which means running the parser groovy code (configured here) directly on the controller which <em>CAN BE UNSAFE</em> from a security point of view.
+  </p>
+    e.g. if your groovy code is poorly written.
```
Is this sentence intended to be in the middle of 2 `<p>` tags?

![Bildschirmfoto 2021-06-29 um 11 11 28](https://user-images.githubusercontent.com/503338/123770801-dc8aa800-d8ca-11eb-85ef-b72fa27ed7ca.png)

pjdarton replied:
🤷 It was deliberate, yes ... and I know that the resulting message renders ok, but I don't know whether or not it would be better to have that middle sentence within its own `<p>` ... `</p>` section too.
i.e. deliberate ... but potentially misguided 😉

uhafner commented on plugin/src/main/java/io/jenkins/plugins/analysis/core/model/ReportScanningTool.java:
```diff
@@ -294,10 +294,29 @@ public FormValidation doCheckPattern(@AncestorInPath final AbstractProject<?, ?>
+        public boolean hasDefaultPattern() {
+            // Maintenance note: We must use the same "is this empty/blank" logic as used in the runtime code.
+            final String defaultPattern = getPattern();
+            return !StringUtils.isBlank(defaultPattern);
```
```suggestion
            return StringUtils.isNotBlank(defaultPattern);
```

uhafner commented on plugin/src/main/resources/io/jenkins/plugins/analysis/core/model/ReportScanningTool/config.jelly:
```diff
@@ -4,15 +4,24 @@
       <f:entry title="${%title.pattern}" field="pattern"
-               description="${%description.overridePattern(descriptor.pattern)}">
+               description="${%description.patternOrDefault('http://ant.apache.org/manual/Types/fileset.html', descriptor.pattern)}"
+               help="${descriptor.getHelpFile('patternOrDefault')}">
```
Good idea! I did not know that this can be changed...

uhafner reviewed (approved):
Thanks for this high quality PR!
